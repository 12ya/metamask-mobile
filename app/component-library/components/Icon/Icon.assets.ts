/* eslint-disable import/prefer-default-export, import/no-commonjs, @typescript-eslint/no-require-imports */
import { AssetByIconName, IconName } from './Icon.types';

/**
 * Asset stored by icon name
 */
export const assetByIconName: AssetByIconName = {
  [IconName.LockFilled]: require('./assets/lock-filled.png'),
  [IconName.AddOutline]: require('./assets/add-outline.png'),
};
