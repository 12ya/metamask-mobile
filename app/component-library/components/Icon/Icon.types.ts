import { ImageSourcePropType, ViewProps } from 'react-native';

/**
 * Icon sizes
 */
export enum IconSize {
  Xss = '10',
  Xs = '12',
  Sm = '16',
  Md = '20',
  Lg = '24',
  Xl = '32',
}

/**
 * Icon names
 */
export enum IconName {
  LockFilled = 'LockFilled',
  AddOutline = 'AddOutline',
}

/**
 * Asset stored by icon name
 */
export type AssetByIconName = {
  [key in IconName]: ImageSourcePropType;
};

/**
 * Icon component props.
 */
export interface IconProps extends ViewProps {
  /**
   * Enum to select between icon sizes.
   */
  size: IconSize;
  /**
   * Enum to select between icon names.
   */
  name: IconName;
}

/**
 * Style sheet input parameters.
 */
export type IconStyleSheetVars = Pick<IconProps, 'size' | 'style'>;
