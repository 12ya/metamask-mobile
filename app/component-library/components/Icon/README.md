# Icon

Icon is a component which is used to render icons from our icon set.

## Props

This component extends `ViewProps` from React Native's [View Component](https://reactnative.dev/docs/view).

### `size`

Enum to select between size variants.

| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :-------------------------------------------------- | :------------------------------------------------------ |
| [BaseAvatarSize](./BaseAvatar.types.ts#L6)          | Yes                                                     |
