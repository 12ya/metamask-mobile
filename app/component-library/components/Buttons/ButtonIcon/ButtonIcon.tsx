/* eslint-disable react/prop-types */

// Third party dependencies.
import React, { useCallback, useState } from 'react';
import { GestureResponderEvent, Pressable } from 'react-native';

// External dependencies.
import Icon from '../../Icons/Icon';
import { useStyles } from '../../../hooks';

// Internal dependencies.
import { ButtonIconProps } from './ButtonIcon.types';
import stylesheet from './ButtonIcon.styles';
import {
  DEFAULT_BUTTONICON_SIZE,
  DEFAULT_BUTTONICON_ICONCOLOR,
  ICONSIZE_BY_BUTTONICONSIZE,
} from './ButtonIcon.constants';

const ButtonIcon = ({
  iconName,
  onPress,
  onPressIn,
  onPressOut,
  style,
  size = DEFAULT_BUTTONICON_SIZE,
  iconColor = DEFAULT_BUTTONICON_ICONCOLOR,
  isDisabled = false,
  ...props
}: ButtonIconProps) => {
  const [pressed, setPressed] = useState(false);
  const { styles } = useStyles(stylesheet, {
    style,
    size,
    pressed,
    isDisabled,
  });

  const triggerOnPressedIn = useCallback(
    (e: GestureResponderEvent) => {
      setPressed(true);
      onPressIn?.(e);
    },
    [setPressed, onPressIn],
  );

  const triggerOnPressedOut = useCallback(
    (e: GestureResponderEvent) => {
      setPressed(false);
      onPressOut?.(e);
    },
    [setPressed, onPressOut],
  );

  return (
    <Pressable
      style={styles.base}
      onPress={!isDisabled ? onPress : undefined}
      onPressIn={!isDisabled ? triggerOnPressedIn : undefined}
      onPressOut={!isDisabled ? triggerOnPressedOut : undefined}
      accessible
      disabled={isDisabled}
      accessibilityRole="button"
      accessibilityState={{ disabled: isDisabled }}
      {...props}
    >
      <Icon
        name={iconName}
        size={ICONSIZE_BY_BUTTONICONSIZE[size]}
        color={iconColor}
        accessibilityLabel={`icon-${iconName}`}
      />
    </Pressable>
  );
};

export default ButtonIcon;
