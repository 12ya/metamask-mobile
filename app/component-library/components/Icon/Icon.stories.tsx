import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { select } from '@storybook/addon-knobs';
import Icon, { IconSize, IconName } from './';

storiesOf('Component Library / Icon', module)
  .addDecorator((getStory) => getStory())
  .add('Default', () => {
    const sizeSelector = select('Size', IconSize, IconSize.Md);

    return <Icon name={IconName.Medal} size={sizeSelector} />;
  });
