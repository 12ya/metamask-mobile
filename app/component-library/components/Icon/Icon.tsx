/* eslint-disable react/prop-types */
import React from 'react';
import { Image } from 'react-native';
import { useStyles } from '../../hooks/useStyles';
import { IconProps } from './Icon.types';
import styleSheet from './Icon.styles';
import SVG from 'react-native-svg';

const Icon = ({ size, style }: IconProps) => {
  const styles = useStyles(styleSheet, {
    size,
    style,
  });

  return <Image height={size} width={size} style={styles.container}></Image>;
};

export default Icon;
