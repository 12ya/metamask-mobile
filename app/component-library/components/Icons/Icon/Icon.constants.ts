/* eslint-disable import/prefer-default-export */
// Internal dependencies.
import { IconName, IconProps } from './Icon.types';

// Sample consts
export const SAMPLE_ICON_PROPS: IconProps = {
  name: IconName.AddOutline,
};
