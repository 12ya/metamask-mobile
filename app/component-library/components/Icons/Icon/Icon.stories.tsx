// Third party dependencies.
import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { color, select } from '@storybook/addon-knobs';

// External dependencies.
import { storybookPropsGroupID } from '../../../constants/storybook.constants';

// Internal dependencies.
import Icon from './Icon';
import { IconProps, IconSize, IconName } from './Icon.types';

export const getIconStoryProps = (): IconProps => ({
  size: select('size', IconSize, IconSize.Md, storybookPropsGroupID),
  name: select('name', IconName, IconName.LockFilled, storybookPropsGroupID),
  color: color('color', '', storybookPropsGroupID),
});
const IconStory = () => <Icon {...getIconStoryProps()} />;

storiesOf('Component Library / Icons', module).add('Icon', IconStory);

export default IconStory;
