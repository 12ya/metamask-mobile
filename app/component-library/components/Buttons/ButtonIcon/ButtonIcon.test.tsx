// Third party dependencies.
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

// External dependencies.
import { IconName } from '../../Icons/Icon';
import { mockTheme } from '../../../../util/theme';

// Internal dependencies.
import ButtonIcon from './ButtonIcon';
import {
  DEFAULT_BUTTONICON_SIZE,
  DEFAULT_BUTTONICON_ICONCOLOR,
  DEFAULT_BUTTONICON_ICONNAME,
} from './ButtonIcon.constants';
import { ButtonIconSizes } from './ButtonIcon.types';

describe('ButtonIcon', () => {
  it('should render correctly', () => {
    const { toJSON } = render(
      <ButtonIcon
        iconColor={DEFAULT_BUTTONICON_ICONCOLOR}
        iconName={DEFAULT_BUTTONICON_ICONNAME}
        size={DEFAULT_BUTTONICON_SIZE}
        onPress={jest.fn}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with default props', () => {
    const { getByLabelText, getByRole } = render(
      <ButtonIcon iconName={DEFAULT_BUTTONICON_ICONNAME} />,
    );
    const icon = getByLabelText(`icon-${DEFAULT_BUTTONICON_ICONNAME}`);
    const button = getByRole('button');
    expect(button).toBeTruthy();
    expect(icon).toBeTruthy();
    expect(icon.props.name).toBe(DEFAULT_BUTTONICON_ICONNAME);
    expect(icon.props.color).toBe(mockTheme.colors.icon.default);
    expect(button.props.style.height.toString()).toBe(DEFAULT_BUTTONICON_SIZE);
  });

  it('should call onPress when pressed', () => {
    const onPressMock = jest.fn();
    const { getByRole } = render(
      <ButtonIcon iconName={IconName.Add} onPress={onPressMock} />,
    );
    const button = getByRole('button');

    fireEvent.press(button);
    expect(onPressMock).toHaveBeenCalled();
  });

  it('should not call onPress when disabled', () => {
    const onPressMock = jest.fn();
    const { getByRole } = render(
      <ButtonIcon iconName={IconName.Add} onPress={onPressMock} isDisabled />,
    );
    const button = getByRole('button');

    fireEvent.press(button);
    expect(onPressMock).not.toHaveBeenCalled();
  });

  it('should call onPressIn and onPressOut', () => {
    const onPressInMock = jest.fn();
    const onPressOutMock = jest.fn();
    const { getByRole } = render(
      <ButtonIcon
        iconName={IconName.Add}
        onPressIn={onPressInMock}
        onPressOut={onPressOutMock}
      />,
    );
    const button = getByRole('button');

    fireEvent(button, 'pressIn');
    expect(onPressInMock).toHaveBeenCalled();

    fireEvent(button, 'pressOut');
    expect(onPressOutMock).toHaveBeenCalled();
  });

  it('should apply custom icon color', () => {
    const { getByLabelText } = render(
      <ButtonIcon iconName={IconName.Add} iconColor="red" />,
    );
    const icon = getByLabelText(`icon-${IconName.Add}`);

    expect(icon.props.color).toBe('red');
  });

  it('should apply correct size', () => {
    const { getByRole } = render(
      <ButtonIcon iconName={IconName.Add} size={ButtonIconSizes.Lg} />,
    );
    const button = getByRole('button');

    expect(button.props.style.height.toString()).toBe(ButtonIconSizes.Lg);
  });

  it('should render different icons based on iconName', () => {
    const { getByLabelText, rerender } = render(
      <ButtonIcon iconName={IconName.Add} />,
    );
    let icon = getByLabelText(`icon-${IconName.Add}`);
    expect(icon).toBeTruthy();

    rerender(<ButtonIcon iconName={IconName.AddSquare} />);
    icon = getByLabelText(`icon-${IconName.AddSquare}`);
    expect(icon).toBeTruthy();
  });

  it('should change backgroundColor when pressed', () => {
    const { getByRole } = render(<ButtonIcon iconName={IconName.Add} />);
    const button = getByRole('button');

    // Check initial background color
    expect(button.props.style.backgroundColor).toBe(undefined);

    // Simulate pressIn
    fireEvent(button, 'pressIn');
    expect(button.props.style.backgroundColor).toBe(
      mockTheme.colors.background.pressed,
    );

    // Simulate pressOut
    fireEvent(button, 'pressOut');
    expect(button.props.style.backgroundColor).toBe(undefined);
  });
});
