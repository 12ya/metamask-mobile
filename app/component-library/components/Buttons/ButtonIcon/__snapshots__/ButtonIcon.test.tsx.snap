// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon should render correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderRadius": 8,
      "height": 24,
      "justifyContent": "center",
      "opacity": 1,
      "width": 24,
    }
  }
>
  <SvgMock
    accessibilityLabel="icon-Add"
    color="#24272A"
    height={16}
    name="Add"
    style={
      {
        "height": 16,
        "width": 16,
      }
    }
    width={16}
  />
</View>
`;
